### Tool Depot

In [1]:
#import neccesary libraries
import pandas as pd
import numpy as np
import math

In [2]:
#We will read the libraries and make the accessible
carrier_rates = pd.read_csv('C:/Users/valde/Desktop/Tool Depot/Datasets/Carrier_rates.csv')
historical_orders = pd.read_csv('C:/Users/valde/Desktop/Tool Depot/Datasets/Historical_orders.csv')
historical_shipments = pd.read_csv('C:/Users/valde/Desktop/Tool Depot/Datasets/Historical_shipments.csv')
product_info = pd.read_csv('C:/Users/valde/Desktop/Tool Depot/Datasets/Product_info.csv')


In [3]:
#check of libraries were read properly
carrier_rates_summary = carrier_rates.head()
historical_orders_summary = historical_orders.head()
historical_shipments_summary = historical_shipments.head()
product_info_summary = product_info.head()
display(carrier_rates_summary.head())
display(historical_orders_summary.head())
display(historical_shipments_summary.head())
display(product_info_summary.head())

Carrier Supplier  Str Nbr      Cost  Unnamed: 4
0       X        A        1   $2,540          NaN
1       X        A        2   $1,640          NaN
2       X        B        1   $1,200          NaN
3       X        B        2   $1,200          NaN
4       Y        A        1    $0.12          NaN

Order Dt  Str Nbr Supplier  Product Id  Order Qty    Product Weight   \
0  8/4/2014        1        A           1      4,338                2.2   
1  8/4/2014        1        A           2      5,006                1.2   
2  8/4/2014        2        A           1      3,600                2.2   
3  8/4/2014        2        A           2      5,173                1.2   
4  8/4/2014        1        B           3      1,890                8.3   

   Order Weight   
0         9,544   
1         6,007   
2         7,920   
3         6,208   
4        15,687

Shipment Date  Str Nbr Supplier   Weight   Unnamed: 4
0      8/4/2014        1        A   15,551          NaN
1      8/4/2014        1        B   15,687          NaN
2      8/4/2014        2        A   14,128          NaN
3      8/4/2014        2        B   21,049          NaN
4     8/11/2014        1        A   16,671          NaN

Product Id Product Desc  Package Qty   Weight  Weight Unit of Measure
0           1       Wrench            1       2.2                     LB
1           2          Saw            1       1.2                     LB
2           3        Drill            1       8.3                     LB
3           4       Hammer            1       2.0                     LB

Everything seems good, we move on to check if data is clean

The following code will check if data is clean for most datasets

In [4]:
#Check for data type before this to make sure the data type is correct
carrier_rates.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Carrier     8 non-null      object 
 1   Supplier    8 non-null      object 
 2   Str Nbr     8 non-null      int64  
 3   Cost        8 non-null      object 
 4   Unnamed: 4  0 non-null      float64
dtypes: float64(1), int64(1), object(3)
memory usage: 452.0+ bytes


In [5]:
#Remove empty column from carriers_rates as its empty
carrier_rates = carrier_rates.drop(columns=['Unnamed: 4'])

In [6]:
#Check again if column is not there anymore
carrier_rates.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Carrier   8 non-null      object
 1   Supplier  8 non-null      object
 2   Str Nbr   8 non-null      int64 
 3   Cost      8 non-null      object
dtypes: int64(1), object(3)
memory usage: 388.0+ bytes


Empty column was removed from dataset

In [7]:
#Check for data type before this to make sure the data type is correct
historical_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 312 entries, 0 to 311
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order Dt          312 non-null    object 
 1   Str Nbr           312 non-null    int64  
 2   Supplier          312 non-null    object 
 3   Product Id        312 non-null    int64  
 4    Order Qty        312 non-null    object 
 5    Product Weight   312 non-null    float64
 6    Order Weight     312 non-null    object 
dtypes: float64(1), int64(2), object(4)
memory usage: 17.2+ KB


In [8]:
#Check for data type before this to make sure the data type is correct
historical_shipments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Shipment Date  208 non-null    object 
 1   Str Nbr        208 non-null    int64  
 2   Supplier       208 non-null    object 
 3    Weight        208 non-null    object 
 4   Unnamed: 4     0 non-null      float64
dtypes: float64(1), int64(1), object(3)
memory usage: 8.3+ KB


In [9]:
#Remove the empty column from datasets historical_shipment
historical_shipments = historical_shipments.drop(columns=['Unnamed: 4'])

In [10]:
#Double check if column has been removed
historical_shipments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Shipment Date  208 non-null    object
 1   Str Nbr        208 non-null    int64 
 2   Supplier       208 non-null    object
 3    Weight        208 non-null    object
dtypes: int64(1), object(3)
memory usage: 6.6+ KB


Empty column has been removed

In [11]:
#Check for data type before this to make sure the data type is correct
product_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Product Id              4 non-null      int64  
 1   Product Desc            4 non-null      object 
 2   Package Qty             4 non-null      int64  
 3    Weight                 4 non-null      float64
 4   Weight Unit of Measure  4 non-null      object 
dtypes: float64(1), int64(2), object(2)
memory usage: 292.0+ bytes


Since we expect the same number of hammers to be sold compare to the wrench we will take this as a reference for the future sales.

### Step 1 and 2

In [12]:
#Create a new dataset with the expect sales for the wrench/hammer as they are expect to sell equally
#Convert the Order date to datetime
historical_orders['Order Dt'] = pd.to_datetime(historical_orders['Order Dt'])

In [13]:
#Filter data the data from "Wrench" product which is 1
wrench_orders = historical_orders[historical_orders['Product Id'] == 1]

In [14]:
#Remove extra space from the columns now allow to find the correct name
wrench_orders.columns = wrench_orders.columns.str.strip()

In [15]:
#Remove any commas from the Order Qty and convert it to integers
wrench_orders['Order Qty'] = wrench_orders['Order Qty'].replace(',', '', regex=True).astype(int)

C:\Users\valde\AppData\Local\Temp\ipykernel_18488\940646906.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wrench_orders['Order Qty'] = wrench_orders['Order Qty'].replace(',', '', regex=True).astype(int)


In [16]:
#Group these by the month for better visualization later on
wrench_orders_per_month = wrench_orders.groupby(wrench_orders['Order Dt'].dt.to_period('M'))['Order Qty'].sum()

In [17]:
#Display the dataset
display(wrench_orders_per_month)

Order Dt
2014-08    28016
2014-09    31047
2014-10    25415
2014-11    28400
2014-12    42091
2015-01    30045
2015-02    29966
2015-03    41180
2015-04    33189
2015-05    36505
2015-06    45874
2015-07    38937
Freq: M, Name: Order Qty, dtype: int32

In [18]:
#Add each months sale for the wrench to know the total number of wrenches sold per the year
total_wrench_per_year = wrench_orders_per_month.sum()
print(total_wrench_per_year)

410665


Total wrenches sold per year 410665

In [19]:
#To measure the total weight for all shipments we will multiple the number of hammers/wrenches by the weight
weight_per_hammer = 2
total_weight_hammer = total_wrench_per_year * weight_per_hammer
print(total_weight_hammer)

821330


Total weight for the hammer shipments is 821330

In [20]:
#Since we expect a increase in sales for hammers/wrenches next year, we will increase the total amount of sold items
total_wrench_per_year_with_increase = total_wrench_per_year * 1.10
print(total_wrench_per_year_with_increase)

451731.50000000006


Total wrenches expected to sale next year 451731.50

In [21]:
#To measure the total weight for all the shipments if sales were to increase by 10%
total_weight_hammer_with_increase = total_wrench_per_year_with_increase * weight_per_hammer
print(total_weight_hammer_with_increase)

903463.0000000001


Total weight after the expected 10% increase in sales 903463.0

In [22]:
#Export for future visualization
#wrench_orders_per_month.to_csv('wrench_orders_per_month.csv', index=False)

Since we expect a 10% increase year over year we will make a new dataset which will show what the sales will look like.

In [23]:
#Strip extra space from column names
historical_orders.columns = historical_orders.columns.str.strip()

In [24]:
#Ensure 'Order Qty' is numeric after cleaning up any commas
historical_orders['Order Qty'] = historical_orders['Order Qty'].replace(',', '', regex=True).astype(int)

In [25]:
#Create a brand new dataset including the 10% increase in sale for all products
historical_orders_with_increase = historical_orders.copy()
historical_orders_with_increase['Order Qty'] = historical_orders_with_increase['Order Qty'] * 1.10

In [26]:
#Display the first rows of the new dataset with the 10% increased in sales
historical_orders_with_increase.head(5)

Order Dt  Str Nbr Supplier  Product Id  Order Qty  Product Weight  \
0 2014-08-04        1        A           1     4771.8             2.2   
1 2014-08-04        1        A           2     5506.6             1.2   
2 2014-08-04        2        A           1     3960.0             2.2   
3 2014-08-04        2        A           2     5690.3             1.2   
4 2014-08-04        1        B           3     2079.0             8.3   

  Order Weight  
0       9,544   
1       6,007   
2       7,920   
3       6,208   
4      15,687

In [27]:
#Group the new dataset by month and calculate the total sales for each item per month
historical_orders_with_increase['Order Dt'] = pd.to_datetime(historical_orders_with_increase['Order Dt'])

In [28]:
#Group by product and month, then sum the 'Order Qty' to get total sales per month for each product
item_sales_per_month_increase = historical_orders_with_increase.groupby([historical_orders_with_increase['Order Dt'].dt.to_period('M'), 'Product Id'])['Order Qty'].sum()

In [29]:
#Display the grouped results
item_sales_per_month_increase.head(10)

Order Dt  Product Id
2014-08   1             30817.6
          2             47254.9
          3             20128.9
2014-09   1             34151.7
          2             60705.7
          3             24179.1
2014-10   1             27956.5
          2             45694.0
          3             19552.5
2014-11   1             31240.0
Name: Order Qty, dtype: float64

In [30]:
#Export for future visualization
#item_sales_per_month_increase.to_csv('item_sales_per_month_increase.csv', index=False)

In [31]:
#Group the new dataset by month and calculate the total sales for each item per month
historical_orders['Order Dt'] = pd.to_datetime(historical_orders['Order Dt'])

In [32]:
#Group by product and month, then sum the 'Order Qty' to get total sales per month for each product
item_sales_per_month = historical_orders.groupby([historical_orders['Order Dt'].dt.to_period('M'), 'Product Id'])['Order Qty'].sum()

In [33]:
#Display the monthly grouped result
item_sales_per_month.head(10)

Order Dt  Product Id
2014-08   1             28016
          2             42959
          3             18299
2014-09   1             31047
          2             55187
          3             21981
2014-10   1             25415
          2             41540
          3             17775
2014-11   1             28400
Name: Order Qty, dtype: int32

In [34]:
#Export for future visualization
#item_sales_per_month.to_csv('item_sales_per_month.csv', index=False)

### Step 3

Flat rate with carrier X calculations for Store 1 and Store 2

In [35]:
#Define the flat rates for Carrier X from the table
flat_rate_store_1_A = 2540  # for Supplier A, Store 1
flat_rate_store_2_A = 1640  # for Supplier A, Store 2
flat_rate_store_1_B = 1200  # for Supplier B, Store 1
flat_rate_store_2_B = 1200  # for Supplier B, Store 2


In [36]:
#Define Carrier X's maximum shipment capacity (in pounds)
shipment_capacity = 44000  # pounds per shipment

In [37]:
#Calculate the total weight of hammers
total_weight = total_wrench_per_year * weight_per_hammer
print(total_weight)

821330


In [38]:
#Split the weight evenly between Store 1 and Store 2
total_weight_store_1 = total_weight / 2
total_weight_store_2 = total_weight / 2
print(total_weight_store_1)
print(total_weight_store_2)

410665.0
410665.0


In [39]:
#Calculate the number of shipments required for each store (rounding up)
shipments_store_1 = math.ceil(total_weight_store_1 / shipment_capacity)
shipments_store_2 = math.ceil(total_weight_store_2 / shipment_capacity)
print(shipments_store_1)
print(shipments_store_2)

10
10


10 shipments are being sent two both stores

In [40]:
#Calculate the total shipping cost for each store for Supplier A
total_shipping_cost_store_1_A = shipments_store_1 * flat_rate_store_1_A
total_shipping_cost_store_2_A = shipments_store_2 * flat_rate_store_2_A
print(total_shipping_cost_store_1_A)
print(total_shipping_cost_store_2_A)

25400
16400


Total cost for each store to get items shipped 
25400 from supplier A to store 1
16400 from supplier A to store 2

In [41]:
#Calculate the total shipping cost for each store for Supplier B
total_shipping_cost_store_1_B = shipments_store_1 * flat_rate_store_1_B
total_shipping_cost_store_2_B = shipments_store_2 * flat_rate_store_2_B
print(total_shipping_cost_store_1_B)
print(total_shipping_cost_store_2_B)

12000
12000


Total cost for each store to get items shipped 
12000 from supplier B to store 1
12000 from supplier B to store 2

In [42]:
#Calculate the total shipping cost for both stores for Supplier A and Supplier B
total_shipping_cost_A = total_shipping_cost_store_1_A + total_shipping_cost_store_2_A
total_shipping_cost_B = total_shipping_cost_store_1_B + total_shipping_cost_store_2_B

In [43]:
#Print the results
print(f"Total shipping cost for Supplier A with Carrier X: ${total_shipping_cost_A}")
print(f"Total shipping cost for Supplier B with Carrier X: ${total_shipping_cost_B}")

Total shipping cost for Supplier A with Carrier X: $41800
Total shipping cost for Supplier B with Carrier X: $24000


Variable rates with carrier Y calculations for Store 1 and Store 2


In [44]:
#Define the per-pound rates for Carrier Y
rate_per_pound_store_1_A = 0.12  # for Supplier A, Store 1
rate_per_pound_store_2_A = 0.07  # for Supplier A, Store 2
rate_per_pound_store_1_B = 0.07  # for Supplier B, Store 1
rate_per_pound_store_2_B = 0.06  # for Supplier B, Store 2

In [45]:
#Calculate the total weight of hammers (split between two stores)
total_weight_store_1 = (total_wrench_per_year / 2) * weight_per_hammer
total_weight_store_2 = (total_wrench_per_year / 2) * weight_per_hammer
print(total_weight_store_1)
print(total_weight_store_2)

410665.0
410665.0


In [46]:
#Calculate the total shipping cost for each store for Supplier A
total_shipping_cost_store_1_A = total_weight_store_1 * rate_per_pound_store_1_A
total_shipping_cost_store_2_A = total_weight_store_2 * rate_per_pound_store_2_A
print(total_shipping_cost_store_1_A)
print(total_shipping_cost_store_2_A)

49279.799999999996
28746.550000000003


Total shipping cost for each store for Supplier
Store 1 Supplier A 49279.79
Store 2 Supplier A 28746.55

In [47]:
#Calculate the total shipping cost for each store for Supplier B
total_shipping_cost_store_1_B = total_weight_store_1 * rate_per_pound_store_1_B
total_shipping_cost_store_2_B = total_weight_store_2 * rate_per_pound_store_2_B
print(total_shipping_cost_store_1_B)
print(total_shipping_cost_store_2_B)

28746.550000000003
24639.899999999998


Total shipping cost for each store for Supplier
Store 1 Supplier B 28746.55
Store 2 Supplier B 24639.89

In [48]:
#Calculate the total shipping cost for both stores for Supplier A and Supplier B
total_shipping_cost_A = total_shipping_cost_store_1_A + total_shipping_cost_store_2_A
total_shipping_cost_B = total_shipping_cost_store_1_B + total_shipping_cost_store_2_B
print(total_shipping_cost_A)
print(total_shipping_cost_B)

78026.35
53386.45


In [49]:
#Print the results
print(f"Total shipping cost for Supplier A with Carrier Y: ${total_shipping_cost_A:.2f}")
print(f"Total shipping cost for Supplier B with Carrier Y: ${total_shipping_cost_B:.2f}")

Total shipping cost for Supplier A with Carrier Y: $78026.35
Total shipping cost for Supplier B with Carrier Y: $53386.45


### Step 4
Increase all outcomes since we expect a 10% increase

In [56]:
#Increase the total weight by 10% since we expect an increase in sales for next year
total_weight_next_year = total_weight * 1.10
print(total_weight_next_year)

903463.0000000001


In [57]:
total_weight_store_1 = total_weight_next_year / 2
total_weight_store_2 = total_weight_next_year / 2
print(total_weight_store_1)
print(total_weight_store_2) 

451731.50000000006
451731.50000000006
